include/mbedtls/mbedtls_config.h

```diff
@@ -3916,4 +3916,9 @@
  */
 //#define MBEDTLS_ECDH_VARIANT_EVEREST_ENABLED
 
+/**
+ * Platform independent implementation for crypto algorithms.
+ */
+//#define MBEDTLS_AES_HAS_NO_BUILTIN /* Uncomment to disable built-in platform independent code of AES */
```

Comment:
This name looks confuse you. This means no Pure-C code. ( platform independent code). 
How about `MBEDTLS_AES_HAS_NO_IND_ACCEL`?

Reply:
I rename it to  `MBEDTLS_AES_HAS_NO_PLAIN_C` , is it better?

Reply:
The problem with `HAS_` is that it sounds like a statement (of fact) rather than an imperative (instruction).

In SHA-2, we've used `MBEDTLS_SHAxxx_USE_A64_CRYPTO_ONLY`, so we could use something like `MBEDTLS_AES_USE_HARDWARE_CRYPTO_ONLY` or `MBEDTLS_AES_DONT_USE_SOFTWARE_CRYPTO`

Reply:
Renamed to `MBEDTLS_AES_DONT_USE_SOFTWARE_CRYPTO` .

Reply:
Sorry to nitpick, but AES-CE and AES-NI are still software IMO. `MBEDTLS_AES_NO_PLAIN_C` is a bit more clear?

Reply:
Or maybe invert it and have `MBEDTLS_AES_PLAINC_C` - this then sits naturally alongside `MBEDTLS_AESCE_C` as one of a possible set of implementations.

Reply:
Yes, that is a good suggestion

Reply:
Having positive variables is easier for users. But it introduces a problem with backward compatibility: if someone has an existing `mbedtls_config.h` with just
```
#define MBEDTLS_AES_C
#define MBEDTLS_AESNI_C
```
then they should get the fallback. We can fiddle with options in `build_info.h`, but that means we'd have to automaticaly enable `MBEDTLS_AES_PLAIN_C` is `MBEDTLS_AES_C` is enabled. Then the user would have to `#undef MBEDTLS_AES_PLAIN_C` in their `mbedtls_config.h` if they don't want fallback. We currently have no pre-enabled configuration options, and I think it would be more confusing to start having one than it would be to have a negative option “no software AES”.

On naming: the suffix `_C` in Mbed TLS compile-time options means “a module providing a feature”. Please avoid option names where it doesn't mean that. Given the precedent of `MBEDTLS_SHA256_USE_A64_CRYPTO_ONLY` and `MBEDTLS_SHA512_USE_A64_CRYPTO_ONLY`, we should use similar names for AES. I suggest

```
MBEDTLS_AES_USE_HARDWARE_ONLY
```
(and really I'd suggest something similar for the SHA) since it makes perfect sense to enable both AESNI and AESCE if you want to deploy both on high-end x86 and high-end arm.

Reply:
I like @gilles-peskine-arm 's suggestion. And it should be negative option to avoid backward compatibility issue.

Reply:
Agreed with @gilles-peskine-arm that we really have to keep it a negative option. But I don't really like "HARDWARE" in the name, because this is a software implementation (that happens to use some special CPU instructions).

So I'm back to `MBEDTLS_AES_NO_PLAINC` as my suggestion. @gilles-peskine-arm suggested (in slack) `MBEDTLS_AES_USE_ACCELERATION_ONLY`. I'm OK with either. @tom-cosgrove-arm WDYT?

Reply:
I've just tried disabling all of the `MBEDTLS_AESCE_C` special CPU features, and also set `MBEDTLS_AES_USE_HARDWARE_ONLY`. It still built and passed the tests, using the plain C implementation, which is not what I expected, given the description says `Disable plain C implementation for AES`.

Should this feature actually disable the plain C? I think it should.

Reply:
Sorry, but after thinking about this for a while, I think it really is HARDWARE.

A software implementation of AES is where code (C, assembler, Rust, JavaScript, whatever) does the calculations using registers and memory on a general-purpose CPU.

Hardware-accelerated AES, like AES-NI and the AArch64 crypto extensions, is where hardware does those calculations - and the CPU is hardware. Yes, software has to set up the input data, but it has to do that when handing off to a separate crypto accelerator.

Therefore I think few people would be confused if we used the term HARDWARE here, but I think some might get confused if we said NO_PLAINC (we're using intrinsics, called by C, so there's a lot riding on the interpretation of "PLAIN" here)

Reply:
Ah, the problem is that the check is in `aesce.c`:

```
#if !defined(MBEDTLS_AESCE_C) && defined(MBEDTLS_AES_USE_HARDWARE_ONLY)
#error "MBEDTLS_AES_C defined, but not all prerequisites"
#endif
```

but... `aes.c` has the following:

```
#if defined(MBEDTLS_AESCE_C)
#include "aesce.h"
#endif
```

so the error will never trigger. Presumably the checks should move into aes.c.